### ISSUE TITLE
[match-case] Incorrect narrowing of types.

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#7823]
match-case does not narrow correctly inside the case clauses. Equivalent if-else code produces different results.

[pyright-playground](https://pyright-play.net/?pyrightVersion=1.1.361&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgElUkRjkAaKAWQIovSwHMAUKEhRUAT14CylarQAK4VGDjx2AFUlIAagRDsCOHE1QB9CRSQGjJ01iQA3JuxABXLBghJTMABYoAaxYkQUEAaSgAXihNS10QAAoAIjCkgEpBbR8wKJiteOSsuCT2OAc9dAIPSPU3JAz1XNidPWT1dNCAATcPcm8-QODBGHhDHCgAZVceORwwpHEcAEEsABMAcSRUBiQIBMUwZVUAbTD2IrAAXTSALkEoB6gQJJeDo4QoYGooMCIAKxQqHGqF8BFoOGmVBoUAABqZTPwtuhGBB4TCoFVVrCAgscDC8C8kqFHlBVkhgFAcYsEsZ4MA0lAALQAPnojGYRGQp0uNygeH590eZIp8MRGBR8JpSDp7CpvLOUAA9AyWVALrz%2BXhQsLPmAwAkAB68qYzGhzXErDZbHYQU4aS5QAA%2BbKYwWOqGmXIV6ku1yZrNW6DQdpiPMFDwgYL8UCN4ZJsEMSEmkNm80Wls22xRCQZhigkZ4fH4d3jpaeW1cICwpKDqASBak-AyZYTxhdHOQOYx42Rrs5SBLLceY1sFiQCV7Has7bdHooXvtfqgAGIoIBQcjjLZAFarNbQE-ZwWbw5sNDsjiYhoygh1RFaRs43EbIZ9Tpn-fdnqQL99KoDtZ-QcIyjXwYyA%2BMYETZNTSBNNljWTMbS7PMGyLcCtx3atA33VCBGPUtILbHYp2QntD37dCyxHM8x0NdhiNnL8fz9VdABlyR9CwEH83wYj85wXUNLk3Mttw9XdsLrScjzjaizHsJxEgNa9bwIAAvQ1eS4Tj%2BG451eK5fjv29X9-T3VBALjdAKXQHA%2BBwVAqhgccDXYE0oVgi0EOtFFbmEh5RMrLDayvSzrNs3AHKwJy6PfZBfJbWTzC0GL9O-QzmIZFd1z88sxKC-clJk085IvRSMiAA)

```python
from collections.abc import Iterable, Mapping
from typing import Protocol, TypeVar, assert_type, assert_never, runtime_checkable

K = TypeVar("K")
V_co = TypeVar("V_co", covariant=True)
T = TypeVar("T")

@runtime_checkable
class SupportsKeysAndGetItem(Protocol[K, V_co]):
    r"""Protocol for objects that support `__getitem__` and `keys`."""

    def keys(self) -> Iterable[K]: ...
    def __getitem__(self, key: K, /) -> V_co: ...

def foo(x: SupportsKeysAndGetItem[K, T] | Iterable[tuple[K, T]]) -> dict[K, T]:
    match x:
        case SupportsKeysAndGetItem() as mapping:
            return dict(mapping)
        case Iterable() as iterable:
            assert_type(iterable, Iterable[tuple[K, T]])  # ✅
            return dict(iterable)
    assert_never(x)

def bar(x: Mapping[K, T] | Iterable[tuple[K, T]]) -> dict[K, T]:
    match x:
        case SupportsKeysAndGetItem() as mapping:
            return dict(mapping)
        case Iterable() as iterable:
            assert_type(x, Iterable[tuple[K, T]]) # ❌ Mapping[K, T] | Iterable[tuple[K, T]]
            return dict(iterable)
    assert_never(x)

def baz(x: Mapping[K, T] | Iterable[tuple[K, T]]) -> dict[K, T]:
    if isinstance(x, SupportsKeysAndGetItem):
        return dict(x)
    if isinstance(x, Iterable):
        assert_type(x, Iterable[tuple[K, T]])  # ✅
        return dict(x)
    assert_never(x)
```

**Comment by randolf-scholz** [comment]
`mypy` produces different results:

```
main.py:20: error: Expression is of type "Iterable[Any]", not "Iterable[tuple[K, T]]"  [assert-type]
main.py:29: error: Expression is of type "Mapping[K, T] | Iterable[tuple[K, T]]", not "Iterable[tuple[K, T]]"  [assert-type]
main.py:37: error: Expression is of type "Mapping[K, T] | Iterable[tuple[K, T]]", not "Iterable[tuple[K, T]]"  [assert-type]
```

[mypy-playgound](https://mypy-play.net/?mypy=latest&python=3.12&gist=1570bbea3192392903b14f32ebb5b4d4)

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.362, which I just published.
